ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
         base_url = self.base_url.replace('http://', 'ws://')
         url = '{}/v1/tasks/{}/websocket/{}'.format(base_url, task_id, websocket_id)
         return url
+
+
+class MultipartLargeFileParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
+        self._buffer = b''
+        self._response = None
+        self._files = {}
+
+        # Store the initial FSM state.  See _run_parser_fsm() for a full explanation.
+        self._initial_state_fn = self._state_detect_boundary
+
+        # Prepare the MIME multipart boundary line matcher.
+        # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]) after the boundary pattern
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        boundary_rgx_pattern = r"--{}(--)?\r\n".format(boundary)
+        self._boundary_rgx = re.compile(boundary_rgx_pattern.encode())
+
+        # When streaming file data to tempfiles, we need to make sure we don't feed part of the
+        # boundary line by mistake.  This is easy if we detect the boundary in full, but if we
+        # haven't received it yet, we need to be careful not to feed any partial boundary line
+        # to disk.  This minimum buffer length threshold will ensure that, and will work for both
+        # regular and terminating boundary cases.
+        #
+        # Calculated as: boundary header length + prefix/suffix -- (4 bytes) + \r\n (2 bytes)
+        self._safe_buffer_len = len(boundary) + 6
+
+    def __del__(self):
+        for file in self._files.values():
+            os.unlink(file.name)
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
```

Comment:
I think that, if we wanted to rely on automagic in Python to be reasonably efficient here, we'd want to use `.join` rather than the `+` operator. That may be dated knowledge -- I'm returning to focusing on Python after a break, myself.

Regardless, that would not address @niemeyer's concerns, because we'd still be making a big, redundant copy of the stuff in memory. Python would just be able to do more of the allocation of the extra memory in advance.

Refactoring to use `self._buffer = data` would be more efficient, because we'd just be changing a reference. That would discard any data left in the buffer, though, which feels intuitively wrong (though might be okay in this case).

This makes me want to set aside some time to build some instrumentation around memory use and such in OF. Without concrete data, it's hard to say how much of an issue the copy is. We are working with big files, but we are working with them in http sized chunks, which is less concerning to me than if we were reading them directly off a local disk.

Reply:
Could we use a `bytearray()` and `self._buffer.extend(data)` to avoid an extra allocation and copy? That way we only need to copy data down when we have a partial buffer, which should only happen once at the start of a large file.

Reply:
Ben is right - we should be able to avoid the reallocation issue by extending a mutable type (e.g. bytearray) rather than using the immutable bytes string.  Doing `self._buffer = bytearray()` in the parser `__init__`  and calling `extend` should do the trick.

Reply:
@rwcarlsen thank you for the comment (and thank you @benhoyt for the other comment, which I managed to miss).

I agree that using a bytearray() is the right thing to do here, with no extra performance testing needed. (This doesn't change my desire to drop a performance testing harness into OF; it just shouldn't block progress on this PR.)